iotile_ext_cloud/iotile/cloud/apps/ota_updater.py

```diff
@@ -0,0 +1,210 @@
+import logging
+import urllib.request
+from datetime import datetime
+import operator
+
+from iotile.core.hw import IOTileApp
+from iotile.core.dev.semver import SemanticVersion
+from iotile.cloud import IOTileCloud, device_id_to_slug
+from typedargs.annotate import docannotate, context
+
+import pytest
+
+op_map = {">=": operator.lt,
+           "gteq": operator.lt,
+           "<=": operator.gt,
+           "lteq": operator.gt,
+           ">": operator.le,
+           "gt": operator.le,
+           "<": operator.ge,
+           "lt": operator.ge,
+           "==": operator.ne,
+           "eq": operator.ne
+          }
+
+
+def _download_ota_script(script_url):
+    """Download the script from the cloud service and store to temporary file location"""
+
+    try:
+        urllib.request.urlretrieve(script_url, "temp.trub")
+        return "temp.trub"
+    except Exception as e:
+        print("Failed to download OTA script")
+        print(e)
+        return False
+
+
+@context("OtaUpdater")
+class OtaUpdater(IOTileApp):
+    """An IOtile app that can get OTA updates from the cloud and apply them to the device.
+    The primary function is a bundled "check_and_update" that
+
+    (1) Checks the cloud for deployment requests
+    (2) Grabs the most recent one for this device (if there are multiple)
+    (3) Checks that the update is applicable (version)
+    (4) Downloads the update script
+    (5) Attempts an update
+    (6) Informs the cloud of results
+
+    Args:
+        hw (HardwareManager): A HardwareManager instance connected to a
+            matching device.
+        app_info (tuple): The app_tag and version of the device we are
+            connected to.
+        os_info (tuple): The os_tag and version of the device we are
+            connected to.
+        device_id (int): The UUID of the device that we are connected to.
+    """
+
+    logger = logging.getLogger(__name__)
+
+    def __init__(self, hw, app_info, os_info, device_id):
+        super(OtaUpdater, self).__init__(hw, app_info, os_info, device_id)
+
+        self.app_info = app_info
+        self.os_info = os_info
+
+        self._cloud = IOTileCloud()
+        self._con = self._hw.get(8, basic=True)
+
+        self.dev_slug = self._get_uuid()
+
+    @classmethod
+    def AppName(cls):
+        return 'ota_updater'
+
+    def _get_uuid(self):
+        device_id, = self._con.rpc(0x10, 0x08, result_format="L16x")
+        return device_id_to_slug(device_id)
+
+    @docannotate
+    def check_and_update(self):
+        """
+        Checks the cloud for an available OTA update and applies it to the device.
+        Informs the cloud of the attempt afterwards
+        Gives some diagnostic info to the user as process goes
+
+        :return: None
+        """
+
+        script = self.check_cloud_for_script()
+        if not script:
+            print("no OTA pending")
+            return
+
+        print("Applying deployment ", script[0])
```

Comment:
Similar to Luke's comment, we try to use `iprint` rather than bare print statements when providing status updates that are primarily relevant to interactive invocation of a method from the `iotile` tool.  `iprint` just calls `print` when run from the `iotile` tool but hides the output when run from a background script.  

This is because we tend to end up (by design) reusing this functionality in many other scripts and bare print statements can't be turned off easily in those contexts, causing unwanted noise on stdout.